ipa-core/src/helpers/stream/chunks.rs

```diff
@@ -72,6 +67,103 @@ impl<K, const N: usize> Chunk<K, N> {
             data: f(data),
         }
     }
+
+    /// Apply a transformation to the chunk data via a `Future`
+    pub fn then<F, Fut, KM>(self, f: F) -> ChunkFuture<Fut, KM, N>
+    where
+        F: FnOnce(K) -> Fut,
+        Fut: Future<Output = Result<KM, Error>>,
+    {
+        let Self { chunk_type, data } = self;
+
+        ChunkFuture::new(f(data), chunk_type)
+    }
+}
+
+enum Expected {
+    /// For the final, partial chunk, we require at least as many sub-chunks as necessary to hold
+    /// the specified amount of data, and at most a full chunk. It is probably only reasonable to
+    /// get one or the other, and not something in between, but we allow that for now.
+    Range(RangeInclusive<usize>),
+    Exactly(usize),
+}
+
+impl Display for Expected {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Expected::Range(bounds) => write!(
+                f,
+                "between {min} and {max}",
+                min = bounds.start(),
+                max = bounds.end()
+            ),
+            Expected::Exactly(value) => value.fmt(f),
+        }
+    }
+}
+
+impl Expected {
+    fn type_str(&self) -> &'static str {
+        match self {
+            Expected::Range(_) => "partial",
+            Expected::Exactly(_) => "full",
+        }
+    }
+
+    fn contains(&self, value: usize) -> bool {
+        match *self {
+            Expected::Range(ref exp_range) => exp_range.contains(&value),
+            Expected::Exactly(exp_value) => exp_value == value,
+        }
+    }
+}
+
+impl<T, const N: usize> Chunk<Vec<T>, N> {
+    /// Unpack nested chunks
+    ///
+    /// This function can be used when converting between vectorization factors. Before calling this function,
+    /// the protocol code needs to convert (e.g. using `Chunk::map` or `Chunk::then`) the chunk data into a
+    /// vector of N / M sub-chunks with data type `T`. This function will then convert the `Chunk<Vec<T>, N>`
+    /// into a vector with type `Vec<Chunk<T, M>>` containing up to N / M chunks. Note that the result may
+    /// be shorter than N / M chunks, if the input was a partial chunk.
+    ///
+    /// # Panics
+    /// If the data stream is not valid for the specified chunk and sub-chunk configuration.
+    #[must_use]
+    pub fn unpack<const M: usize>(self) -> Vec<Chunk<T, M>> {
+        let Self { chunk_type, data } = self;
+        debug_assert!(N % M == 0);
+        let (mut len, expected) = if let ChunkType::Partial(len) = chunk_type {
+            (len, Expected::Range(((len + M - 1) / M)..=(N / M)))
```

Comment:
It would be more work to adjust the last chunk, than to just do the same transformation consistently. If anything, I would prefer to make this require `N / M` chunks in all cases.

If we wanted to adjust the last chunk, note that we can't do it within the vectorized protocol (because the protocol doesn't know how many of the records are valid). It would need to be done between [lines 313 and 314 in `compute_prf_for_inputs`](https://github.com/private-attribution/ipa/pull/1087/files#diff-88f3ba117d83fecf9cf14f0f2148b93845c047164b3770b747c8c04d784b8c08R313-R314).